[RxSwift] JuiceMaker

horeng2 opened the pull request:
# [RxSwift] JuiceMaker

@jsim27 
PR 제출합니다 👍

# 사용 기술
- ### RxSwift
- ### MVVM 패턴
- ### Input / OutPut 모델링

# 구현 내용
### 1. 스토리보드를 사용한 UI 구현
- 스토리보드를 통해 UI를 구현했으며, 각 뷰는 내비게이션컨트롤러의 Push 방식으로 이동합니다.
### 2. View 구성
- 주스를 주문하는 `OrderView`와 과일 재고를 관리하는 `EditView`로 이루어져 있습니다.
### 3. ViewController
- 각 ViewController는 View 로드 시 `binding`메소드를 실행하며, View를 구성할 데이터를 저장합니다.
- `Button` 또는 `Stepper`를 클릭 시 `IBAction`메소드를 통해서 동작합니다.
- output되는 정보에 따라서 `Alert`을 띄웁니다.
### 4. ViewModel
- Input / OutPut 모델링으로 구현했습니다.
- `transform` 메소드를 통해서 기능에 맞는 동작을 실행하고, ViewController에 `Output`을 전달합니다.
### 5. Model
- `fruitRepository`를 가집니다.
- `makeJuice`메소드는 `canMake`메소드를 통해서 과일의 재고가 충분한지 확인하고, 충분하다면 `takeFruitStock`메소드를 통해서 과일의 재고를 차감한뒤 `Observable<Juice>`를 반환합니다. 충분하지 않은 경우 nil을 반환합니다. 

### 6. Repository
- `fruitStock` 프로퍼티가 과일 재고 데이터를 갖습니다.
- `initialStock`, `maximumStock` 프로퍼티를 통해서 재고의 초기값과 최댓값을 가집니다.
- 초기화 시 열거형 `Fruit`과 `initailStock`을 이용해서 초기 데이터를 `fruitStock`에 저장합니다.
- 값을 읽어오고, 수정하는 기능을 구현했습니다. 

# 구현 화면
## 주문 화면
![order](https://user-images.githubusercontent.com/87305744/170487792-fea95a26-d49a-433e-8fcd-3392c293cd7e.gif)
## 재고 최솟값
![minimum](https://user-images.githubusercontent.com/87305744/170487804-47067c52-1f44-4b0b-bf26-7b2d45f99bff.gif)
## 재고 최댓값
![maximum](https://user-images.githubusercontent.com/87305744/170487822-b37a4bd5-269b-49d0-96ed-0d13067415c9.gif)


# 고민한 점
### EditViewController의 Stepper Value
초반 로직은 `ViewModel`에서 `stepper`의 `value`를 현재의 재고값과 함께 계산하여 데이터를 업데이트 하는 방식이었는데,
로직이 복잡해져서 가독성이 떨어지고, 오류 수정도 힘들었습니다. 또한 현재의 재고값을 반환하는 메소드는 `Observable<Int>` 타입을 반환하고 있기 때문에 `Int`타입으로 계산에 이용하려면 클로져의 중첩이 많이 발생했습니다. 데이터와 별도로 매번 계산을 실행하는 과정도 조금 비효율적으로 느껴져서 개선안을 고민했습니다.
-> `stepper`의 `value`가 과일 재고 데이터값과 다른 것이 원인이라 판단하고, `EditView`를 로드할 때 `stepper`의 `value`에 현재 재고값을 로드했습니다. `updateCurrentValue`메소드에서 `Lable`들의 `text` 로드와 함께 실행됩니다.
또한 이 메소드에서 `stepper`의 최솟값과 최댓값을 지정했습니다. 초반 구성에서는 스토리보드를 이용했지만, `FruitRepository`에서 `static`프로퍼티로 최대값을 선언하고 이를 사용해서 변경에 용이하도록 했습니다.

### JuiceMaker의 makeJuice 메소드, canMake 메소드
BehaviorSubject에 레시피에 필요한 과일의 사용가능 여부를 Bool값으로 입력했는데, 제일 최종으로 입력된 하나의 값만 사용되는 문제가 있었습니다. 구독의 시점이 문제인 것 같다고 생각해서 순서에 조금 더 신경써보고, 후에는 RelaySubject 사용도 시도했지만 성공하지는 못했습니다... 
차선책으로 `canMake`라는 별도의 메소드를 작성하여 필요한 과일들의 사용 가능 여부가 담긴 `Observable<[Bool]>`을 반환하도록했습니다. 재고 여부가 하나라도 false라면 주스 제작이 불가능하기 때문에, 이를 이용해서 제작과 실패를 구현했습니다.

# 궁금한 점
### Rx의 Output
Rx를 사용해보는 것이 처음이라서 Output에 대한 감을 잡는데까지 오래걸렸습니다. Input은 입력하는 정보인 것이 이해가 되는데, Output으로는 아무리 생각해도 데이터를 전달하는 방향으로밖에 생각이 안났습니다...😂 그래도 프로젝트를 진행하며 여러 삽질과 공부를 하다보니 아주 쪼오끔은 느낌이 오는 것 같기도 한데요...데이터의 전달과 로직 실행은 Observer, Observeble, 구독 등을 통해서 실시간으로 이루어지고, input과 output은 그것들을 준비하고 구독하거나 트리거 시키는 과정에 필요한 것이라고 정리하면 될까요?

---
로직을 완성하고 실행하는 것만해도 꽤 오래걸려서 정작 깊게 고민하고 궁금했던 점이 별로 없네요😂
그래도 첫 걸음을 내딛는 기분으로 많이 배웠습니다.
구석구석 아낌없는 조언 부탁드려요!!!
나무 최고💖

horeng2 committed:
feat: FruitStore->FruitRepository 리네이밍, readStock 메소드 구현

horeng2 committed:
feat: JuiceMaker 타입 선언 및 메소드 구현

horeng2 committed:
refactor: 뷰모델 및 뷰컨트롤러 로직 변경

horeng2 committed:
feat: EditViewController 클래스 생성 및 내비게이션 화면 전환 구현

horeng2 committed:
refactor: 주문 성공시 재고가 줄지 않는 오류 수정 및 Output에 resultMessage 프로퍼티 추가

horeng2 committed:
feat: 주문 결과 Alert 기능 구현 및 재고 차감 로직 오류 수정

horeng2 committed:
feat: JuiceMaker 타입 내부 modifyFruitStock 타입 구현

horeng2 committed:
feat: EditViewModel 타입 구현

horeng2 committed:
feat: EditViewController Stepper IBOutlet 및 IBAction 추가, 구현

horeng2 committed:
refactor: JuiceMaker 리팩토링으로 쥬스 주문 오류 수정

horeng2 committed:
feat: 재고 수정 기능 구현 완료

horeng2 committed:
feat: 최소/최고 재고를 넘어갈 때 Alert이 발생하는 기능 구현, stepper가 현재 재고값을 가지도록 리팩토링

horeng2 committed:
refactor: stepper의 최소/최대값을View 로드 시 지정하여 해당 값 도달 시 Alert이 뜨도록 리팩토링

horeng2 committed:
refactor: alertMessage를 전달하는 방식 수정

horeng2 committed:
refactor: 과일 재고를 업데이트하는 메소드 리네이밍

horeng2 committed:
chore: 오타 수정

horeng2 committed:
refactor: 접근제어자 설정 및 파일명 변경

horeng2 committed:
chore: 줄공백 및 코딩컨벤션 수정

jsim27 commented on JuiceMaker/JuiceMaker/ViewController/OrderViewController.swift:
```diff
@@ -16,52 +16,101 @@ class OrderViewController: UIViewController {
+
+    private var disposeBag = DisposeBag()
+
     override func viewWillAppear(_ animated: Bool) {
         super.viewWillAppear(animated)
+        self.binding()
     }
     
     override func viewDidLoad() {
         super.viewDidLoad()
-        self.binding()
     }
```
#2  Lifecycle 메서드의 순서를 맞춰주는 것이 어떨까요?!

horeng2 replied:
Life Cycle의 순서에 맞게 메소드 순서 변경했습니다.

jsim27 commented on JuiceMaker/JuiceMaker/ViewController/OrderViewController.swift:
```diff
@@ -16,52 +16,101 @@ class OrderViewController: UIViewController {
+            }).disposed(by: disposeBag)
+    }
+    
+    @IBAction func orderOfStrawberryBananaJuice(_ sender: Any) {
+        self.input.orderJuice.onNext(.strawberryBananaJuice)
+    }
+    
+    @IBAction func orderOfMangoKiwiJuice(_ sender: Any) {
+        self.input.orderJuice.onNext(.mangoKiwiJuice)
+    }
+    
+    @IBAction func orderOStrawberryJuice(_ sender: Any) {
+        self.input.orderJuice.onNext(.strawberryJuice)
+    }
+    
+    @IBAction func orderOfBananaJuice(_ sender: Any) {
+        self.input.orderJuice.onNext(.bananaJuice)
+    }
+    
+    @IBAction func orderOfPineappleJuice(_ sender: Any) {
+        self.input.orderJuice.onNext(.pineappleJuice)
+    }
+    
+    @IBAction func orderOfKiwiJuice(_ sender: Any) {
+        self.input.orderJuice.onNext(.kiwiJuice)
+    }
+    
+    @IBAction func orderOfMangoJuice(_ sender: Any) {
+        self.input.orderJuice.onNext(.mangoJuice)
+    }
```
#3  만약 스토리보드(Interface Builder)를 사용하지 않고 뷰를 그리는 환경이라면 유저 이벤트를 받는 방식을 동일하게 유지할 수 있을까요? 
IBAction을 사용하지 않는다면 어떻게 Input binding을 해줄 수 있을까요?

horeng2 replied:
```swift
let input = orderJuiceModel.Input(
            orderJice: self.orderOfStrawberryBananaJuice.rx.tap.asObservable()
        )
```
초기에 위와 같은 방식으로 binding을 시도했다가, 스토리보드 방식에 맞추어 IBAction을 활용하는 방법으로 변경했던 부분입니다.
이처럼 binding을 한다면 주스주문 버튼 tap에 해당하는 동작을 각각의 input프로퍼티로 따로 binding 합니다.

당시에는 각각의 input 프로퍼티를 가지는 것보다 하나의 프로퍼티로(Juice타입 Subject) 바인딩하는 것이 더 좋다고 생각했는데 지금 짚어주신 부분을 다시 보니, 최초로 바인딩하는 과정이기 때문에 버튼 액션으로 분기하는 것보다는 각각 따로 바인딩하는 것이 rx프로그래밍 방식에 더 적합한 것이 아닌가 하는 생각이 들어요🤔

jsim27 replied:
사실 의미 상 / 기능 상 같기 때문에 크게 문제는 없지만,
1. `binding` 함수 내에서 **모든 Input과 Output을 명세서처럼 모아서 관리할 수 있다**는 점
2. 뷰 구현 방식이 코드베이스로 변하더라도 뷰컨트롤러의 변경점이 적어진다는 점

정도의 장점이 있는 것 같아 말씀드렸던 부분이에요!
수정 확인했습니다 👍

jsim27 commented on JuiceMaker/JuiceMaker/ViewController/OrderViewController.swift:
```diff
@@ -16,52 +16,101 @@ class OrderViewController: UIViewController {
+            }).disposed(by: disposeBag)
+    }
+    
+    private func updateStockLabel(of fruit: Fruit) {
+        var label: UILabel?
+        switch fruit {
+        case .strawberry:
+            label = self.strawberryStockLabel
+        case .banana:
+            label = self.bananaStockLabel
+        case .pineapple:
+            label = self.pineappleStockLabel
+        case .kiwi:
+            label = self.kiwiStockLabel
+        case .mango:
+            label = self.mangoStockLabel
+        }
```
#2 
-  옵셔널로 선언하지 않아도 문제가 없는 것 같아요!
- 혹은 이런 방법은 어떨까요
```swift
let label: UILabel = {
    switch fruit {
    case .strawberry:
        return self.strawberryStockLabel
    case .banana:
        return self.bananaStockLabel
    case .pineapple:
        return self.pineappleStockLabel
    case .kiwi:
        return self.kiwiStockLabel
    case .mango:
        return self.mangoStockLabel
    }
}()
```

horeng2 replied:
이런 경우 연산프로퍼티를 잘 활용해야겠네요! 수정했습니다👍

jsim27 commented on JuiceMaker/JuiceMaker/ViewController/OrderViewController.swift:
```diff
@@ -16,52 +16,101 @@ class OrderViewController: UIViewController {
+    
+    @IBAction func orderOfMangoJuice(_ sender: Any) {
+        self.input.orderJuice.onNext(.mangoJuice)
+    }
```
#1 #3 더불어, 이번 스텝에서는 정형화되고 일반적으로 사용되는 뷰 바인딩 방식을 익혀보기 위해, 다음과 같은 방식으로 리팩토링 해 보길 권장합니다!
- IBAction이 아닌 IBOutlet 사용
```swift
private lazy var input = OrderViewModel.Input(orderJuice: PublishSubject<Juice>())
private lazy var output = orderViewModel.transform(input: input)
```
- 위 부분을 지금처럼 뷰컨트롤러 프로퍼티로 갖고 있는 것이 아닌, binding()함수 내부에서 지역변수로 생성하도록 구현
- IBOutlet UIButton의 `.rx` 프로퍼티(Reactive 익스텐션)에서 제공하는 ControlEvent 활용하여 스트림 시작점으로 사용하기

! 참고 레퍼런스: https://github.com/sergdort/CleanArchitectureRxSwift/tree/master/CleanArchitectureRxSwift/Scenes

horeng2 replied:
해당사항 리팩토링 완료했습니다.🥳

jsim27 commented on JuiceMaker/JuiceMaker/ViewController/OrderViewController.swift:
```diff
@@ -16,52 +16,101 @@ class OrderViewController: UIViewController {
-            .disposed(by: self.disposeBag)
-            
-        
+        alertController.addAction(UIAlertAction(title: "OK", style: .default))
+        present(alertController, animated: true)
+    }
```
#2 
`self.present`
:)

horeng2 replied:
`self` 키워드가 누락된 부분 모두 추가했습니다👍

jsim27 commented on JuiceMaker/JuiceMaker/ViewController/OrderViewController.swift:
```diff
@@ -16,52 +16,101 @@ class OrderViewController: UIViewController {
+            .map{ String($0) }
+            .subscribe(onNext: { stock in
+                label?.text = stock
+            }).disposed(by: disposeBag)
```
1.  #3  
subscribe 대신에 `bind(to:)`를 사용하여 뷰 컴포넌트의 Binder에 연결해 주는 것은 어떨까요?
2. #2
`.disposed` 이전에도 개행을 해주면 어떨까요 :)

horeng2 replied:
subscribe 대신 bind를 사용하여 바인딩하도록 수정했습니다. 덕분에 두 방법의 차이점을 자세히 공부해봤어요.👍
개행 수정했습니다!

jsim27 commented on JuiceMaker/JuiceMaker/ViewController/OrderViewController.swift:
```diff
@@ -16,52 +16,101 @@ class OrderViewController: UIViewController {
+            .subscribe(onNext: { stock in
+                label?.text = stock
+            }).disposed(by: disposeBag)
+    }
```
`updateStockLabel`에서 현재 self.viewModel에 접근하여 `fruitStockObservable(of:)`메서드를 사용해 주고 있는 모습인데요.

제 생각에는 VC의 label에 표시될 정보들 또한 뷰모델에서 내보내 줄 Output에 해당된다고 봐요. 
사실 뷰에 표시될 모든 데이터가 Output이죠. 그래서, 뷰모델의 transform 이외의 메서드를 이용할 필요가 없는 것이 맞다고 봅니다. **transform을 통해 모든 Output 스트림을 뷰모델로부터 가져올 수 있어야** 하니까요.

뷰컨트롤러가 뷰모델의 transform 이외의 메서드를 사용하지 않는 방식으로 리팩토링 해보길 권장합니다. #1 #4 #5 

더 자세한 내용은 뷰모델 파일에서 코멘트 할게요. 역시, 이 부분도 레퍼런스를 적극 참고해 주세요!
https://github.com/sergdort/CleanArchitectureRxSwift/tree/master/CleanArchitectureRxSwift/Scenes

jsim27 replied:
수정 확인했습니다! 👍

jsim27 commented on JuiceMaker/JuiceMaker/ViewController/OrderViewController.swift:
```diff
@@ -16,52 +16,101 @@ class OrderViewController: UIViewController {
-//            orderOfKiwiButtonTap: orderOfKiwiJuice.rx.tap.asObservable(),
-//            orderOfMangoButtonTap: orderOfMangoJuice.rx.tap.asObservable()
+        self.output.orderSuccess
+            .subscribe(onNext: { _ in
+                Fruit.allCases.forEach { fruit in
+                    self.updateStockLabel(of: fruit)
+                }
+            }).disposed(by: disposeBag)
+        self.output.resultMessage
+            .subscribe(onNext: { message in
+                self.showAlert(title: "주문 결과", message: message)
+            }).disposed(by: disposeBag)
```
#2  .disposed 앞에서도 개행해주면 예쁘지 않을까요? 🌷🌷

horeng2 replied:
모두 개행했습니다!🌸

jsim27 commented on JuiceMaker/JuiceMaker/ViewModel/OrderViewModel.swift:
```diff
@@ -7,56 +7,44 @@
+class OrderViewModel {
+    private let juiceMaker = JuiceMaker()
+    private let disposeBag = DisposeBag()
 
     struct Input {
-        let viewWillAppear: Observable<Void>
```
viewWillAppear 인풋 스트림을 `초기 재고값을 label에 표시해 줄 trigger`로 삼았으면 참 좋았을 것 같다고 생각하는데, 왜 지우게 되었을까요?🥲  #4

horeng2 replied:
삭제 당시에는 bind 메소드를 viewWillAppear에서 호출하기 때문에 필요 없다고 생각해서 삭제했었는데, 바인딩 방식을 바꾸면서 원래대로 viewWillAppear도 인풋에 포함하였습니다!

jsim27 commented on JuiceMaker/JuiceMaker/ViewModel/OrderViewModel.swift:
```diff
@@ -7,56 +7,44 @@
-//        let orderOfPineappleButtonTap: Observable<Void>
-//        let orderOfKiwiButtonTap: Observable<Void>
-//        let orderOfMangoButtonTap: Observable<Void>
+        let orderJuice: PublishSubject<Juice>
```
Input 스트림을 `Observable`이나 `Driver`가 아닌, 읽기와 쓰기가 모두 가능한 `Subject` 형태로 전달하고자 한 의도가 무엇인가요?

제 생각에 뷰모델 입장에서 Input은 받아들여서 가공만 해서 Output에 전달해 주면 되는 정보이므로 추가적인 쓰기는 필요가 없을 것 같은데, 이렇게 구현해야 했던 이유를 알려주세요! #4 #5

horeng2 replied:
`Subject`와 `Observable`의 차이에 대한 이해가 부족해서 특별한 이유 없이 `Subject` 형태로 전달했던 것 같습니다. 추가적인 공부와 블로그 글 작성을 통해서 차이점과 적절한 사용에 대해서 이해한 뒤에 `Observable` 형태로 변경했습니다.🤩

jsim27 commented on JuiceMaker/JuiceMaker/ViewModel/OrderViewModel.swift:
```diff
@@ -7,56 +7,44 @@
-//        var mangoStock: Driver<Int>
+        let orderSuccess: BehaviorSubject<Bool>
+        let resultMessage: PublishSubject<String>
     }
```
Output 스트림 역시, Observable이나 Driver가 아닌, 읽기와 쓰기가 모두 가능한 Subject 형태로 전달하고자 한 의도가 무엇인가요?

뷰컨트롤러는 뷰모델에서 전달된 데이터를 받아서 뷰 컴포넌트에 바인딩만 해 주면 되는데, 이 스트림에 쓰기를 할 여지를 열어줄 필요가 없지 않을까요?

이렇게 구현해야 했던 이유를 알려주세요 :) 

#4 #5

horeng2 replied:
위의 코멘트와 마찬가지로 처음에는 개념에 대한 이해가 부족해서 `Subject`를 사용했었는데요, 현재의 코드에서도 `Subject`를 사용하고 있습니다. 그 이유는 모든 버튼 이벤트가 발생할 때마다 재고 업데이트 명령을 하는 `orderButtonBind` 프로퍼티에서 `merge`를 통해 `output`의 이벤트들을 바인딩하고 있기 때문입니다.

jsim27 replied:
transform 과정에서 구현상 쓰기가 필요하여 `Subject`를 사용하게 되더라도, 뷰컨트롤러에 반환할 때는 `asObservable()`을 통해 `Observable`타입으로 변환하여 반환하는게 바람직해 보여요. 뷰컨트롤러에서 의도치 않은 `onNext`를 할 여지를 막고, 그럴 수 있다는 오해를 막기 위해서라도요!
`private(set)` 접근제어를 거는 이유와 비슷하게 생각해도 될 것 같아요!

또한, 모든 버튼 이벤트가 발생할 때마다 재고 업데이트 명령을 하는 로직을 구현할 때에도 뷰모델에서 `bind`(`subscribe`) 및 `dispose`하지 않도록 operator를 사용하여 구현할 수 있을 것 같아요. 

계속 할수있다고만 코멘트하고 어떻게 하는지 지침을 주지 못하는것 같아서, 제가 별도의 예시 코드를 작성해 보았는데, 이건 따로 말씀드릴게요!

jsim27 commented on JuiceMaker/JuiceMaker/ViewModel/OrderViewModel.swift:
```diff
@@ -7,56 +7,44 @@
     
     func transform(input: Input) -> Output {
-        let currentStock = fruitStock.asDriver()
+        let orderSuccess = BehaviorSubject<Bool>(value: true)
+        let resultMessage = PublishSubject<String>()
+        
+        input.orderJuice
+            .map{ juice in
+            self.juiceMaker.makeJuice(juice)
+            }.subscribe(onNext: { juiceObservable in
+                juiceObservable
+                    .subscribe(onNext: { juice in
+                        if juice == nil {
+                            orderSuccess.onNext(false)
+                            resultMessage.onNext(OrderResult.orderFailure.message)
+                        } else {
+                            orderSuccess.onNext(true)
+                            resultMessage.onNext(OrderResult.orderSuccess.message)
+                        }
+                    }).disposed(by: self.disposeBag)
+            }).disposed(by: disposeBag)
```
#4 #5 
transform 메서드는 input 스트림을 받아서 "가공" 하여 output으로 "변형" 해 주는 함수입니다.
하지만, 현재 모습은 단순히 가공(map 등의 operator)하는 것이 아닌, "소비"(subscribe)하고, "생성"(line 25, 26)하고 있네요.

이 부분이 rx-MVVM에서 가장 중요하고, 또 어려운 부분인 것 같아요.
(유저 인풋 이벤트로부터 뷰의 변경까지 이어지는) `"하나의 스트림은 단 번만 .subscribe(혹은 bind, drive) 되어야 한다"` 
이 원칙을 지키도록 노력해 보세요. 그래서, 사실 **뷰모델에서는 .subscribe가 전혀 없는 것이 원칙적으로 올바른 모습**입니다.

![MVVM_StreamFlow 001](https://user-images.githubusercontent.com/72993238/170825822-50d5ff08-1c45-4ce9-b1da-9f3fc0d29467.jpeg)

사실 이 그림은 호랭이의 뷰모델을 보고 설명을 위해 만들기 시작한 그림이에용..
Stream 시작점으로부터 최종 소비되기 전까지는 가공만 이루어지는 것이고, 소비되지 않는 것이 옳다고 생각합니다.

map 대신에 flatMap을 사용하고, subscribe 대신에 map을 사용하면 구독 없이도 충분히 operator만으로 input을 output으로 변형 가능할거에요.

가장 중요한 부분이니만큼 이 부분을 필수적으로 해결하고 머지하도록 할게요 :) #1 
잘 안되면 언제든지 도움을 요청해 주세요!!

horeng2 replied:
Stream에 대한 이해 부족이었던 것 같습니다.😂 자세한 설명 덕분에 잘 이해가 됐습니다! 스트림의 중간과정에서는 최대한 subscribe를 사용하지 않고 operator를 사용하는 방향으로 리팩토링 했습니다.

jsim27 replied:
전보다 아주 많이 개선된 점 확인했습니다! 잘했어용 👏👏👏

jsim27 commented on JuiceMaker/JuiceMaker/ViewController/OrderViewController.swift:
```diff
@@ -16,52 +16,101 @@ class OrderViewController: UIViewController {
-    var disposeBag = DisposeBag()
-    
+    private let orderViewModel = OrderViewModel()
+    private lazy var input = OrderViewModel.Input(orderJuice: PublishSubject<Juice>())
+    private lazy var output = orderViewModel.transform(input: input)
+
+    private var disposeBag = DisposeBag()
+
     override func viewWillAppear(_ animated: Bool) {
         super.viewWillAppear(animated)
+        self.binding()
```
#3  
`self.binding()`을 `viewDidLoad`가 아닌 `viewWillAppear`에서 호출한 의도가 뭔지 궁금해요!
바인딩의 시점은 언제가 되면 좋을까요?

현재 20, 21 line에서 input과 output을 프로퍼티로 가지고 있는 점 + 바인딩 함수의 호출 시점
두 가지 문제가 복합적인 원인이 되어서 다음과 같은 오작동이 있습니다.

- 재고 수정 화면으로 이동했다가 복귀했을 때, 주문 버튼을 누르면 요청이 중복으로 두 번 동작합니다.
- 재고 수정 화면으로 n번 이동했다가 복귀하기를 반복하면, 주문 버튼을 한 번 눌러도 요청이 n번 반복됩니다.

재고가 여러번 깎이지는 않지만, Alert present 로직이 중복 실행되어 콘솔창에 아래와 같은 메시지가 여러 번 출력되고 있어요!


<img width="920" alt="2022-05-28_22-05-40" src="https://user-images.githubusercontent.com/72993238/170826845-67a79031-69d6-4834-935f-1a2fe72e4e75.png">

horeng2 replied:
input에 viewWillAppear를 빼는 과정에서 아예 바인딩을 viewWillAppear 호출 시 실행하면 될 것이라고 생각했는데요, 두 방법의 차이가 아직 완벽히 이해는 안가요🧐 뷰모델에서 input.viewWillAppear를 통해서 선언되는 output 프로퍼티가 존재하지 않고, 트리거 역할만 한다고 생각해서 input 프로퍼티로 viewWillAppear를 넣지 않고, 대신 바인딩 메소드를 viewWillAppear에서 실행해도 같은 결과라고 생각했거든요...제가 놓치고 있는 게 있을까요?

jsim27 replied:
viewWillAppear 될때마다 바인딩메서드가 중복으로 호출되고 있기때문에 다음 화면으로 이동하고 다시 돌아오면 바인딩이 2중, 3중으로 걸려있게 된 것 같아요. 그래서 위 에러가 발생하는 것이구요.
바인딩은 최초 한 번만 연결되고, 그 후에는 바인딩된 방식대로 이벤트에 대한 reaction만 일어나면 되는 것일텐데, 여러번 실행될 여지가 있는 viewWillAppear에 바인딩메서드가 있는 것은 위험해보입니다. 호랭이 말대로 viewWillAppear 트리거만 봤을때는 맞는 말이 될 수 있는데, 그 외의 스트림들이 다중 연결되며 의도치 않은 동작이 위 에러상황처럼 일어나게 되니까요!

jsim27 replied:
이 부분에 대한 의문은 해소되신걸까요? @horeng2

horeng2 replied:
넵 덕분에요!👍

1. viewDidLoad에서 바인딩 할 경우
-> view가 로드 된 뒤 한번만 바인딩이 일어나고, 그 이후로 input 프로퍼티로 등록된 트리거 viewWillApear가 호출될 때에는 최초 바인딩으로 연결된 것을 통해서 리액션이 실행된다.
2. viewWillAppear에서 바인딩 할 경우
-> viewWillApear가 될 때마다 바인딩이 일어난다. 바인딩이 다중으로 여러번 실행된다.

- 요점: 바인딩은 최초 한번만 실행되고, viewWillApear는 이벤트 발생 트리거로 이용된다.

위와 같이 이해했습니다.

바인딩을 통한 연결과 이벤트 발생에 대한 이해가 부족해서 의문점이 있었습니다.
뷰가 로드될 때 최초로 바인딩을 하면 이후에는 바인딩 필요성이 없는데,
저 의문을 가질 때에는 bind를 실행하는 시점에서 데이터를 불러오는 것이 실행된다고 생각했거든요.
UIKit을 사용할 때 메소드 호출을 통해서 반환값을 사용했던 것과 혼동이 있었던 것 같아요.

차이점에 대해 전혀 감이 잡히지 않아서 의문 투성이로 코멘트를 남겼던 게 불과 4일 전이네요! 
나무 덕분에 많이 성장했어요.🥹

jsim27 commented on JuiceMaker/JuiceMaker/Repository/FruitRepository.swift:
```diff
@@ -0,0 +1,43 @@
+
+import Foundation
+import RxSwift
+import RxOptional
```
#2 
RxOptional 호랭이 덕분에 알게 되었어요ㅎㅎ 정말 편리해 보이네요!
그런데, 기본 Rx의 Operator에 익숙해지기 위해서 이번 프로젝트에서는 해당 라이브러리를 배제하고 기본 Operator만을 이용해서 동일한 기능을 다시 구현해보는것은 어떨까요..? 조심스레 제안해 봅니다!

jsim27 commented on JuiceMaker/JuiceMaker/ViewController/EditViewController.swift:
```diff
@@ -0,0 +1,111 @@
+
+    private let disposeBag = DisposeBag()
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        binding()
+    }
```
1. `self.` :) #2 
2. 이전 뷰컨과는 달리 viewDidLoad에서 binding()을 호출한 의도가 무엇인가요오? #3

horeng2 replied:
1. 수정했습니다!
2. 원래는 여기서도 viewWillAppear에서 호출할 생각이었는데 단순 누락이었습니다...😅

jsim27 replied:
위 viewWillAppear에 대한 코멘트를 참고해서 binding 메서드의 위치를 수정해보면 좋을 것 같아요:)

jsim27 commented on JuiceMaker/JuiceMaker/ViewController/EditViewController.swift:
```diff
@@ -0,0 +1,111 @@
+            }).disposed(by: disposeBag)
+        
+        self.output.alertMessage
+            .subscribe(onNext: { message in
+                guard let message = message else {
+                    return
+                }
+                self.showAlert(title: "알림", message: message)
+            }).disposed(by: disposeBag)
+    }
+    
+    private func updateCurrentValue(of fruit: Fruit) {
+        var label: UILabel?
+        var stepper: UIStepper?
+        switch fruit {
+        case .strawberry:
+            label = self.strawberryStockLabel
+            stepper = self.strawberryStepper
+        case .banana:
+            label = self.bananaStockLabel
+            stepper = self.bananaStepper
+        case .pineapple:
+            label = self.pineappleStockLabel
+            stepper = self.pineappleStepper
+        case .kiwi:
+            label = self.kiwiStockLabel
+            stepper = self.kiwiStepper
+        case .mango:
+            label = self.mangoStockLabel
+            stepper = self.mangoStepper
+        }
+        stepper?.minimumValue = .zero
+        stepper?.maximumValue = Double(FruitRepository.maximumStock)
+        
+        editViewModel.fruitStockObservable(of: fruit)
+            .subscribe(onNext: { stock in
+                label?.text = String(stock)
+                stepper?.value = Double(stock)
+            }).disposed(by: disposeBag)
+    }
```
이전 뷰컨 파일에서 코멘트한 것과 동일한 이슈들은 여기서는 생략하도록 할게요.

### 1. 다만 생각해 볼 부분은, 
```swift
Fruit.allCases.forEach { fruit in
    self.updateCurrentValue(of: fruit)
}
```
이렇게 순회하며 과일의 종류에 따라 적합한 레이블과 스테퍼를 매칭시켜주는 로직을 갖기보다는, 
아예 Output이 각 label과 stepper별로 세부적으로 나뉘어서 제공되면 어떨까 해요. 뷰컨은 어떤 판단을 하거나 로직을 갖기보다, 주어진 정보를 있는 그대로 연결시켜주기만 한다는 MVVM의 패턴에 걸맞게 말이죵.. #4 #3 

### 2. 또한, 
```swift
stepper?.minimumValue = .zero
stepper?.maximumValue = Double(FruitRepository.maximumStock)
```
이 부분은 updateCurrentValue가 아니라, stepper들을 configure해 주는 로직 아닐까요? 별개의 메서드에 있는게 적합해 보여요! #2

horeng2 replied:
`configure` 메소드 분리했습니다!

jsim27 commented on JuiceMaker/JuiceMaker/ViewController/EditViewController.swift:
```diff
@@ -0,0 +1,111 @@
+        alertController.addAction(UIAlertAction(title: "OK", style: .default))
+        present(alertController, animated: true)
+    }
+}
```
이전 뷰컨에서 말씀드린 부분을 생각해 주시고, 변경 부탁드립니다! #1 #3 #4

jsim27 replied:
수정 확인했습니다! 👏👏👏👏

jsim27 commented on JuiceMaker/JuiceMaker/ViewModel/EditViewModel.swift:
```diff
@@ -0,0 +1,59 @@
+import Foundation
+import RxSwift
+
+class EditViewModel {
```
OrderViewModel 파일과 거의 동일한 구조로 만들어져 있으므로, 해당 파일의 코멘트를 참고해 주세용!! #1  #4  #5

jsim27 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -0,0 +1,54 @@
+    
+    func makeJuice(_ juice: Juice) -> Observable<Juice?> {
+        let juiceObservable = self.canMake(juice: juice)
+            .map{ $0.contains(false) ? false : true }
```
#2 리빙포인트
```swift
.map { $0.contains(false) }
.map(!)
```
이런 것도 되더라구요.. `!`도 함수라서 가능한..!!

horeng2 replied:
오오! 결과를 반전시켜주는군용ㅎㅎㅎㅎ꿀팁 감사합니다👍

jsim27 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -0,0 +1,54 @@
+    func makeJuice(_ juice: Juice) -> Observable<Juice?> {
+        let juiceObservable = self.canMake(juice: juice)
+            .map{ $0.contains(false) ? false : true }
+            .map{ $0 == true ? juice : nil }
```
`.map { $0 ? juice : nil }`
도 가능하지만 가독성 때문에 위와 같이 표현하신걸까요? #2

horeng2 replied:
네! 다른 사람이 내 코드를 본다고 생각했을 때, 현재값이 참이라면?이라는 조건을 보자마자 직관적으로 이해할 수 있도록 풀어서 써봤습니다.🧐 map과 $0은 보통 참거짓 판별보다는 데이터의 가공에서 많이 쓰인다고 생각해서 $0만 작성한다면 삼항연산의 뒷부분까지 읽거나 한번 더 생각을 해야할 것 같아서요..!!

jsim27 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -0,0 +1,54 @@
+    private let diposeBag = DisposeBag()
+    
+    func fruitStockObservable(of fruit: Fruit) -> Observable<Int> {
+        return fruitRepository.readStock(of: fruit)
```
#2 `self.`😉

jsim27 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -0,0 +1,54 @@
+                    return
+                }
+                self.takeFruitStock(for: juice)
+            })
```
do 오퍼레이터의 활용 멋져요!!
개인적으로 정말 중요하고, 잘 다루면 유용한 오퍼레이터라고 생각합니다ㅎㅎ

jsim27 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -0,0 +1,54 @@
+        juice.recipe.forEach { (fruit, count) in
+            fruitRepository.readStock(of: fruit)
+                .map{ $0 >= count }
+                .subscribe(onNext: { canMake.append($0) })
+                .disposed(by: diposeBag)
```
본 JuiceMaker 객체는 Clean아키텍쳐로 말하자면 도메인 레이어, 즉 UseCase에 해당하는 부분인데요. 비즈니스 로직이 있는 부분이지요.
뷰에서 어떤 이벤트 Input이 들어오면, 뷰모델에서 UseCase를 '사용'하여 Input stream을 가공 및 합성해 주게 됩니다.

일반적으로 UseCase는 Repository로부터 어떤 모델(Juice 혹은 재고) 데이터의 Stream을 요청하여 적절한 가공만 거친 후 그 Stream을 그대로 뷰모델에 return해 주지요.
하지만 여기서는 subscribe를 통해 `소비`하여 변수에 저장하고, Observable로 한 번 감싸(`생성`) 리턴해 주는 모습이네요.

OrderViewModel에서 남긴 코멘트와 같은 맥락에서, UseCase는 스트림이 소비되기 적합한 곳이 아닙니다..!

잠깐 정리를 해 보자면, 
- Stream의 생성: ViewController, Repository
- Stream의 가공: ViewModel, UseCase
- Stream의 소비: ViewController

이렇게 정리할 수 있겠네요!

그러므로, .subscribe는 VC에서만 하는 원칙을 지키도록 해 보세요 :) #1 #6

horeng2 replied:
소비, 가공, 생성의 흐름에 대해서 다시한번 생각해보고, subscribe를 통해 소비하는 시점에 대해 유의하며 리팩토링 했습니다.👍

jsim27 commented on JuiceMaker/JuiceMaker/Repository/FruitRepository.swift:
```diff
@@ -0,0 +1,43 @@
+
+class FruitRepository {
+    static let shared = FruitRepository()
+    private(set) var fruitStock = [Fruit : Int]()
```
그냥 `private`이 되어도 문제 없어 보이네요!  #2

horeng2 replied:
불필요한 set 키워드 삭제했습니다👍

jsim27 commented on JuiceMaker/JuiceMaker/Repository/FruitRepository.swift:
```diff
@@ -0,0 +1,43 @@
+    
+    init() {
+        Fruit.allCases.forEach { fruit in
+            fruitStock.updateValue(self.initialStock, forKey: fruit)
```
`self.`🍰 #2

jsim27 commented on JuiceMaker/JuiceMaker/Repository/FruitRepository.swift:
```diff
@@ -0,0 +1,43 @@
+    
+    func readStock(of fruit: Fruit) -> Observable<Int> {
+        return Observable.just(self.fruitStock[fruit])
+                .filterNil()
```
- `self.fruitStock[fruit, default: ]` 딕셔너리의 이 서브스크립트 문법을 사용하거나
- filterNil 대신에 compactMap을 사용할 수도 있겠네요.
#2

jsim27 commented on JuiceMaker/JuiceMaker/Repository/FruitRepository.swift:
```diff
@@ -0,0 +1,43 @@
+
+class FruitRepository {
+    static let shared = FruitRepository()
+    private(set) var fruitStock = [Fruit : Int]()
```
!! 이 코멘트는 바로 반영하지 말고 따로 의논해봤으면 좋겠어요..!!



fruitStock 이라는 데이터소스를 변수가 아닌 BehaviorSubject / PublishSubject로 변경하면 구현이 훨씬 쉬울 것 같아요!
이 부분은 저희가 미리 공지를 했어야 하는데 미흡했던 것 같습니다😭 

현재 호랭이는 매번 모델을 변경한 후에, 모든 과일 재고를 read하는 방식으로 뷰를 바인딩하고 있지만
데이터소스 자체를 옵저버블 스트림으로 뷰에 연결해 놓으면 매번 read하는 과정을 생략할 수 있을 것 같아요!
#6

horeng2 replied:
저번에 의논한 것을 바탕으로 Observable로 변경했습니다. 하나의 딕셔너리 요소만을 가지고 있고, 다른 Observable 객체를 바라볼 필요도 없다는 판단으로 Subject가 아닌 Observable로 구현했는데요, 만약 Subject로 구현할 경우에 제가 생각하지 못한 장점이나 현재 구조에서 제가 놓친 부분이 있을까요?

jsim27 replied:
이실직고하자면, Subject로 고치자고 한 것은 저와 기획자들의 판단미스였던 것 같아요.. 죄송합니다ㅜ
Subject와 Relay를 헷갈려서, Relay처럼 현재 값을 빼와서 수정 후에 onNext할 수 있는 줄로만 착각했던 거였어요..

그래서 기존대로 [Fruit : Int] 변수로도 충분할 것 같고, 혹은 호랭이가 현재 바꾼 것처럼 Observable도 괜찮은 것 같습니다!

jsim27 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -0,0 +1,54 @@
+    func makeJuice(_ juice: Juice) -> Observable<Juice?> {
+        let juiceObservable = self.canMake(juice: juice)
+            .map{ $0.contains(false) ? false : true }
+            .map{ $0 == true ? juice : nil }
```
!!이 코멘트는 반영하지 말고 의논해보는게 좋겠어요. onError를 보내는게 꼭 정답은 아닐것같다는 생각도 들어서요..!!



`self.canMake`로 제조 가능한 재고가 있는지를 판단한 후, 재조가 불가능하다면 `nil`을 내보내는 방식으로 로직을 구성하셨는데요!

nil은 비어있음 / 없을 수 있음을 뜻하는 개념인 만큼, 에러나 불가능을 나타내기 위해 사용하는 것은 의미 상 혼동을 줄 수 있을 것 같아요.
repository에 요청하는 makeJuice에 대한 return Obeservable stream 에서부터 onError로 에러를 보내주는 것이 어떨까요?

전체적인 로직 수정이 불가피하겠지만, 아주 중요한 부분인 것 같으니 고민을 해 보았으면 좋겠어요! #6

horeng2 replied:
이부분에 대해 고민을 해봤는데요,
아래의 두가지 문제를 가지고 있다고 판단하고 이를 바탕으로 리팩토링을 진행했습니다.
1. 의미없는 nil값이 쓰인다.
2. map을 통해서 어떤 동작을 하고 어떤 결과를 얻어내고자 하는지 명확하지 않다.

$0 대신 매개변수명을 통해 명확하게 뜻을 나타낼 수 있도록 변경하였고, 반환타입을 Observable<Bool> 타입으로 변경하여 주스 제작의 성공 여부를 반환하도록 했습니다.
관련된 `OrderViewModel`의 코드도 함께 변경되었는데, 보다 뜻이 명확하고 직관적인 코드로 개선된 것 같습니다!

jsim27 reviewed (commented):
@horeng2 정말 고생많았습니다! 🐯

제가 RxSwift를 처음 적용할 때는, 숲재와 둘이서 하기도 했고 오랜 기간을 투자해서 겨우 구현이 가능했는데, 호랭이는 짧은 기간 공부해서 도움을 받지 않고 기능 구현을 했다는 것 자체가 너무 대단한 것 같아요.. Rx천재!!

이벤트 발생부터 스트림이 어디로 흘러가는지 "방향" 은 전부 다 정확해서, 어떻게 이렇게 잘 했는지 처음 보고 정말 놀랐어요!

상세 코멘트에서 언급하는 부분이지만, "방향" 은 맞지만 "방식" 의 수정은 조금 필요할 것 같아요!

현재 VC -> VM -> UseCase -> Repository -> UseCase -> VM -> VC 로 흐르는 스트림의 거의 모든 구간에서 subscribe를 하고, 새롭게 생성한 Observable(Subject)에 onNext로 값을 저장한 후, 그것을 다음 구간으로 전달하는 방식으로 뚝뚝 끊기고 있어요. 고속도로로 치면 매 커브마다 톨게이트가 있다고나 할까요...

이것을 하나의 이어진 길로 쭉~~ 이어줄 수 있도록 리팩토링해야 하는데, 쉽지 않을거에요.. 
필수적으로 리팩토링해야 하는 부분에는 #1 태그를 달아 두었습니다!

구체적인 가이드라인이나 힌트를 코멘트에 적어놓았는데, 다시한번 참고자료를 언급하자면!
- 우리가 바이블처럼 참고해야하는 [Repository 링크](https://github.com/sergdort/CleanArchitectureRxSwift/tree/master/CleanArchitectureRxSwift/Scenes) 와
- ![MVVM_StreamFlow 001](https://user-images.githubusercontent.com/72993238/170831556-01a77a2b-dd10-4a2e-81b0-555632e37a19.jpeg)
요 그림을 참고해 주세요ㅎㅎ


코멘트들을 고민해보시고, 호랭이가 주도하고 제가 보조하는 식으로 라이브 리팩토링을 해 봐도 좋을 것 같고

궁금한거나 같이 고민해보고 싶은 점이 있으면 DM & 코멘트 & 디스코드 언제든 편하게 알려주세용

마지막으로 다시한번 강조하고 싶은건,, 코멘트들은 개선점 위주로만 남겼을 뿐이고 진짜진짜 잘하셨어요!!! 못다한 칭찬을 여기에 남기면서 리뷰 마무리합니당ㅎㅎ 💯 💯 💯

jsim27 commented:
# 궁금한 점
> Rx의 Output
Rx를 사용해보는 것이 처음이라서 Output에 대한 감을 잡는데까지 오래걸렸습니다. Input은 입력하는 정보인 것이 이해가 되는데, Output으로는 아무리 생각해도 데이터를 전달하는 방향으로밖에 생각이 안났습니다...😂 그래도 프로젝트를 진행하며 여러 삽질과 공부를 하다보니 아주 쪼오끔은 느낌이 오는 것 같기도 한데요...데이터의 전달과 로직 실행은 Observer, Observeble, 구독 등을 통해서 실시간으로 이루어지고, input과 output은 그것들을 준비하고 구독하거나 트리거 시키는 과정에 필요한 것이라고 정리하면 될까요?

## Comment
Input은 호랭이 생각대로 View에서 발생하는 모든 이벤트(ViewController LifeCycle 메서드, Button tap 이벤트를 비롯한 사용자 이벤트 등)를 의미하고, Output은 `View 에 표시되어야 할 모든 데이터`를 의미합니다! 

위에 첨부한 그림을 뜯어보시면 더 이해가 쉬울 것 같아요...!

1. 모든 input 이벤트를 트리거 삼아서 뷰모델의 로직이 실행되고, 
2. 뷰모델에서는 해당 input들의 operator에서 적절한 UseCase(본 프로젝트에서는 JuiceMaker)의 메서드를 사용해서 Repository에 있는 모델 데이터 변경을 요청하거나 데이터의 stream을 반환받습니다!
3. 그 후 뷰에 표시되어야 할 모습대로 위 stream을 가공하거나 합치거나 나누어서 Output으로 전달해 주는 것이지요!

이 모든 것이 ViewModel의 transform 함수 내에서 이루어지는 것이구요!

horeng2 committed:
chore: 메소드 순서 변경 #2

horeng2 committed:
refactor: updateCurrentValue 내부 프로퍼티를 연산프로퍼티로 변경 #2

horeng2 committed:
chore: 누락된 self 키워드 추가 #2

horeng2 committed:
chore: 코딩컨벤션에 맞추어 개행 수정 #2

- disposed 개행

horeng2 committed:
feat: configure 메소드 구현 #2

horeng2 committed:
chore: RxOptional 삭제 및 해당 코드 수정 #2

horeng2 committed:
chore: 불필요한 set 키워드 삭제 #2

horeng2 committed:
refactor: OrderView binding 방식 및 초기값 세팅 방식 변경 #3

horeng2 committed:
refactor: fruitStock을 dictionary에서 Observable로 리팩토링 #6

horeng2 committed:
refactor: FruitRepository의 fruitStock을 Observable에서 Subject로 변경 #6

horeng2 committed:
refactor: JuiceMaker 타입 내부 fruitStockObservable의 반환 타입을 Observable<String>으로 변경 #6

horeng2 committed:
refactor: OrderViewModel 구조 전반적으로 리팩토링#4 #5
- Input 타입에 viewWillAppear 프로퍼티 추가
- Output 타입 orderSuccess 프로퍼티 삭제, orderButtonBind 프로퍼티 추가
- Output 타입에 과일재고별로 각각의 Subject 프로퍼티 선언
- 버튼을 누를 때마다 재고감소, 알람 표시가 실행되고 input에 해당하는 동작들을 merge하여 input발생 시 과일 재고가 업데이트 되도록 구현

horeng2 committed:
refactor: drive->bind로 변경 #3

horeng2 committed:
chore: 공백줄 제거, 코딩컨벤션에 맞추어 수정 #2

horeng2 committed:
chore: 사용하지 않는 메소드 삭제, 테스트용 임시 코드 수정 #2

horeng2 committed:
refactor: FruitRepository 내부 재고 수정 메소드 오류 수정 #6

horeng2 committed:
refactor: updateFruitStock 메소드 반환값 없도록 리팩토링 #6

horeng2 committed:
refactor: updateFruitStock 메소드 반환값 없도록 리팩토링 #6

horeng2 committed:
refactor: 반환타입 변경 #6

horeng2 committed:
refactor: 반환 타입 변경으로 인한 코드 수정 #6

horeng2 committed:
refactor: Input, Ouput 타입 및 transform 구조 변경 #4 #5

horeng2 committed:
refactor: 기존의 IBAction을 통한 바인딩에서 IBOutlet을 이용한 바인딩 방식으로 변경 #3

horeng2 committed:
chore: 사용되지 않는 IBAction 연결 해제 #2

horeng2 committed:
chore: orderViewModel 프로퍼티의 위치를 binding 메소드 내부로 이동 #3

horeng2 committed:
refactor: makeJuice 메소드 로직 변경 및 직관적인 클로저 매개변수명으로 변경 #6

horeng2 committed:
chore: guard문의 조건에서 == true 키워드 삭제 #3

horeng2 commented:
좋은 리뷰, 세심한 코멘트 덕분에 코드가 많이 개선되었습니다.
코멘트 주신 내용을 참고하여 여러 부분을 수정했습니다!
아직 해결못한 부분이 한가지 있어서 리팩토링은 미완이지만 PR 후 시간이 꽤 지나버려서...여기서 피드백 사항 보고하겠습니다.🤓

문제의 부분은 아래와 같습니다.

`editViewModel`의 `transform`메소드에서 
```swift
        let strawberryStock = input.strawberryStepperDidTap
            .map{ stepperValue in
                let newValue = Int(stepperValue)
                alertMessage.onNext(self.limitStockAlertMessage(newValue))
                self.juiceMaker.updateFruitStock(for: .strawberry, newQuantity: newValue)
            }
            .flatMap{ self.juiceMaker.fruitStockObservable(of: .strawberry) }
```
바로 이 부분인데요, 바인딩 할 시점의 stepperValue는 0이기 떄문에 fruitStock에 0값이 들어가서 재고가 모두 날아가버리는 현상😂입니다.
바인딩하기 전에 stepper의 value를 현재 재고값과 같게 설정해주고 싶어서....여러가지 로직과 방법을 시도중입니다.

이부분 조언부탁드리며...😭 다시한번 감사인사드려요!!!!!
나무 짱짱!!우주최강!!!🤩

horeng2 committed:
chore: 메소드 호출 누락 및 개행 수정 #3

horeng2 committed:
feat: 초기값 세팅을 위한 Output 프로퍼티 currentStock 선언 #3 #5

horeng2 commented:
윗글에서 이야기한 문제를 해결해서 추가글 작성합니다!

1. Output에 현재의 재고값을 담는 딕셔너리 타입의 프로퍼티 `currentStock`을 추가
2. transform에서 현재 재고값을 `currentStock` 프로퍼티에 담아 `Output`을 통해서 [Fruit: Double] 형태로 뷰에 전달
3. 뷰컨트롤러의 `binding` 메소드에서 각각의 과일별로 바인딩하기 전에 `currentStock`을 이용해서 초기값 세팅

현재의 값을 [Fruit:Double]형태로 전달하기 위해서 EditViewModel에서 `subscribe`를 사용했는데, 
뷰에 필요한 값을 위해 뷰모델에서 Observable을 소비하고 가공해서 넘겨주는 것이므로 소비 시점에 문제가 없다고 판단했습니다.
제 생각이 맞을까요?
다른 좋은 방안이 있거나, 제 판단에 틀린 부분이 있다면 조언 부탁드립니다.🤓

horeng2 committed:
refactor: 메소드명 변경 #2

horeng2 committed:
chore: 사용하지 않는 메소드 삭제 #2

horeng2 committed:
chore: 누락된 self 키워드 추가, 공백줄 제거, 사용하지 않는 프로퍼티 삭제 #2

jsim27 commented:
## Comment
>  바인딩 할 시점의 stepperValue는 0이기 떄문에 fruitStock에 0값이 들어가서 재고가 모두 날아가버리는 현상😂

### a. 해결 방법
사실 `skip(1)` 오퍼레이터를 각 스텝퍼 인풋에 걸어버렸으면 쉽게 해결할 수 있었을 건데
제가 코멘트 확인이 늦는 바람에 호랭이가 더 공부할 수 있는 기회가 되었다고 생각할게요.. 용서해주세요😭
각 스텝퍼에서 방출하는 최초값인 0을 거르고 그 다음 이벤트부터 받도록 하면 되는 부분이었던 것 같아요!

### b. 위의 방법과 별개로, 수정하신 방향에 대한 제 의견을 말씀드리면
이런 식의 가공이 불가피했던 이유는, 재고를 읽어오는 방식이 **각 과일별로 분리되어있기 때문** 인 것 같아요.
어차피 Repository에서도 전체 재고를 읽되 `Observable.just(stocks[fruit])` 와 같이 fruit 기준으로 값을 한번 더 추출해줄 뿐인데요
이 방식대로라면 전체 재고를 읽어와야 할 때 Fruit.forEach를 돌아야 하고, repository 메서드도 그만큼 여러번 호출해야 하는 불편함이 생기는 것 같아요. 
즉, **read메서드를 전체 재고를 딕셔너리형태로 읽어오는 방식으로 고친다**면 쉽게 해결되겠죠..


### c. 굳이 repository메서드를 고치지 않고도 해결할 방법을 찾자면,
다른 stream처럼 일관성있게 `Observable<[Fruit: Double]>` 형태로 가공하여 output으로 반환할 방법이 있긴 해요!
```swift
Observable.zip(
    Fruit.allCases.map { fruit in
        self.juiceMaker.fruitStockObservable(of: fruit)
            .map { (fruit, Double($0)) }
    }
)
.map(Dictionary<Fruit, Double>.init)
```
1. 과일별로 재고 스트림을 가져와서 (Fruit, Double) 튜플로 map // 현재 타입은 `[Observable[(Fruit, Double)]]`
2. 위 스트림들이 들어있는 배열을 zip // 현재 타입은 `Observable<[(Fruit, Double)]>`
3. `2`를 `Dictionary`의 `.init(uniqueKeysWithValues:)` 이니셜라이저를 통해 딕셔너리로 변환
-> 결과는 `Observable<[Fruit: Double]>`
이제 이것을 output으로 내보내 뷰컨에서 subscribe해주면 되겠죵..!


---
여러 방법을 말씀드렸지만, a, b, c 순으로 추천합니다.


뷰모델은 아무리 생각해 봐도 bind나 subscribe가 있을 곳이 아닌 것 같아요.
물론 시간의 제약 속에서 구현이 급하다면 유연하게 허용해도 되겠지만, 우리는 공부하는 입장이니만큼 원칙에 충실한 방식으로 하면 좋을 것 같아요!

jsim27 commented:
@horeng2 

수정사항 모두 검토했고, 코멘트 답변 완료했습니다 ^^7

전보다 Operator의 적극적인 사용을 통해 불필요한 구독/dispose가 거의 사라진 것이 눈에 띄었어요! 
정말 까다로운 부분인데 고생 많으셨습니다!!
이제 Rx에 많이 익숙해지신것 같아요🎉🎉

그리고, [PR에 대한 참고용 예시 코드](https://github.com/horeng2/RxSwift_JuiceMaker/pull/7#issue-1263830450)도 별도의 Draft PR로 남겨 보았어요.

검토해 보시고, 원하시는 추가적인 수정을 마치면 다시 알려주세요! 
그 때 Merge하겠습니당ㅎㅎ

horeng2 commented:
수정 완료해서 코멘트 남깁니다!

남겨주신 코멘트를 참고해서 먼저 저 나름대로 코드를 작성해 본 뒤에, 남겨주신 참고용 코드를 살펴보고 필요한 부분을 추가 수정 및 commit 하는 방식으로 리팩토링 진행했습니다.
예시 코드 덕분에 효율적인 방식으로 가독성있고 깔끔하게 작성하는 것에 많은 도움이 되었어요!
디테일이나 코드 속 숨은 꿀팁으로도 많이 배웠습니다.

코멘트를 보고 제가 생각한 리팩토링 방향과 예시 코드의 방향이 대부분 일치해서
프로젝트 진행 기간동안 제 자신이 꽤 성장했구나 싶어서 뿌듯하기도 했습니다.😊

정말 감사합니다. 나무는 최고의 리뷰어예요! 🌳💖
